File: qMRI/R/IRmodels.r
IRhomogen <- function(par, InvTimes){
  ##
  ## Inversion Recovery MRI  1 compartment (used for CSF)
  ##
  ## S_{InvTime} = par[1] * abs( 1 - 2 * exp(-InvTime*par[2]) )
  ##
  n <- length(InvTimes)
  z <- .Fortran(C_irfluid,
                as.double(par),
                as.double(InvTimes),
                as.integer(n),
                fval = double(n),
                grad = double(2*n))[c("fval", "grad")]
  fval <- z$fval
  attr(fval,"gradient") <- matrix(z$grad, n, 2)
  fval
}

IRmix2 <- function(par, InvTimes, S0f, Rf){
  ##
  ## Inversion Recovery MRI 2 compartments (fluid/solid) with fixed fluid parameters
  ##
  ## S_{InvTime} = abs(S0f *  par[1] * (1-2*exp(-InvTime*Rf)) +
  ##                        (1-par[1])*par[3]* (1-2*exp(-InvTime*par[2])) )
  ##
  n <- length(InvTimes)
  z <- .Fortran(C_irmix,
                as.double(par),
                as.double(InvTimes),
                as.double(S0f),
                as.double(Rf),
                as.integer(n),
                fval = double(n),
                grad = double(3*n))[c("fval", "grad")]
  fval <- z$fval
  attr(fval, "gradient") <- matrix(z$grad, n, 3)
  fval
}

IRmix2fv <- function(par, InvTimes, S0f, Rf){
  ##
  ## Inversion Recovery MRI 2 compartments (fluid/solid) with fixed fluid parameters
  ##
  ## S_{InvTime} = abs(S0f *  par[1] * (1-2*exp(-InvTime*Rf)) +
  ##                        (1-par[1])*par[3]* (1-2*exp(-InvTime*par[2])) )
  ##
  n <- length(InvTimes)
  z <- .Fortran(C_irmixfv,
                as.double(par),
                as.double(InvTimes),
                as.double(S0f),
                as.double(Rf),
                as.integer(n),
                fval = double(n))[c("fval")]
  fval <- z$fval
  fval
}

IRmix5 <- function(par, InvTimes){
  ##
  ## Inversion Recovery MRI 2 compartments (fluid/solid) with fixed fluid parameters
  ##
  ## S_{InvTime} = abs(th[5] *  par[1] * (1-2*exp(-InvTime*th[4])) +
  ##                        (1-par[1])*par[3]* (1-2*exp(-InvTime*par[2])) )
  ##
  n <- length(InvTimes)
  z <- .Fortran(C_irmix5,
                as.double(par),
                as.double(InvTimes),
                as.integer(n),
                fval = double(n),
                grad = double(5*n))[c("fval", "grad")]
  fval <- z$fval
  attr(fval, "gradient") <- matrix(z$grad, n, 5)
  fval
}

IRmix5fv <- function(par, InvTimes){
  ##
  ## Inversion Recovery MRI 2 compartments (fluid/solid) with fixed fluid parameters
  ##
  ## S_{InvTime} = abs(th[5] *  par[1] * (1-2*exp(-InvTime*th[4])) +
  ##                        (1-par[1])*par[3]* (1-2*exp(-InvTime*par[2])) )
  ##
  n <- length(InvTimes)
  z <- .Fortran(C_irmix5fv,
                as.double(par),
                as.double(InvTimes),
                as.integer(n),
                fval = double(n))[c("fval")]
  fval <- z$fval
  fval
}

LSIRmix2 <- function(par, Y, InvTimes, S0f, Rf){
   fval <- IRmix2fv(par, InvTimes, S0f, Rf)
   kval <- (Y - fval) 
   value <- sum(kval^2)
   value
}

LSIRmix2grad <- function(par, Y, InvTimes, S0f, Rf){
  fval <- IRmix2(par, InvTimes, S0f, Rf)
  grad <- attr(fval,"gradient")
  kval <- (Y - fval) 
  gradient <- -2 * kval %*% grad
  gradient
}

LSIRmix5 <- function(par, Y, InvTimes){
  fval <- IRmix5fv(par, InvTimes)
  kval <- (Y - fval) 
  value <- sum(kval^2)
  value
}

LSIRmix5grad <- function(par, Y, InvTimes){
  fval <- IRmix5(par, InvTimes)
  grad <- attr(fval,"gradient")
  kval <- (Y - fval) 
  gradient <- -2 * kval %*% grad
  gradient
}

IRmix2fix <- function(par, InvTimes, S0f, S0s, Rf, Rs){
  ##
  ## Inversion Recovery MRI 2 compartments (fluid/solid) with fixed fluid and solid parameters
  ##              mixture parameter only
  ##
  ## S_{InvTime} = abs(S0f *  par[1] * (1-2*exp(-InvTime*Rf)) +
  ##                        (1-par[1])*par[3]* (1-2*exp(-InvTime*par[2])) )
  ##
  n <- length(InvTimes)
  z <- .Fortran(C_irmix0,
                as.double(par),
                as.double(InvTimes),
                as.double(Rs),
                as.double(S0s),
                as.double(S0f),
                as.double(Rf),
                as.integer(n),
                fval = double(n),
                grad = double(n))[c("fval", "grad")]
  fval <- z$fval
  attr(fval, "gradient") <- matrix(z$grad, n, 1)
  fval
}

IRhomogenQL <- function(par, InvTimes, CL, sig, L){
  ##
  ## Inversion Recovery MRI  1 compartment (used for CSF)
  ##
  ## S_{InvTime} = par[1] * abs( 1 - 2 * exp(-InvTime*par[2]) )
  ##
  n <- length(InvTimes)
  z <- .Fortran(C_irfluid,
                as.double(par),
                as.double(InvTimes),
                as.integer(n),
                fval = double(n),
                grad = double(2*n))[c("fval", "grad")]
  sfval <- z$fval/sig
  fval <- CL * hg1f1(-.5, L, -sfval*sfval/2)
  CC <- CL * hg1f1(.5, L+1, -sfval*sfval/2) * sfval/2/L/sig
  grad <- matrix(CC*z$grad, n, 2)
  ind <- sfval>100
  if(any(is.na(ind))){
    warning(paste("IRhomogenQL\n","par",par,"sigma",sig,
                  "\n fv",z$fval,"\n CC",CC,"ind",ind),call.=FALSE)
  }
  if(any(ind)){
    # use LS if there is no real difference, factor to keep monotonicity
    fval[ind] <- z$fval[ind]*1.0001
    grad[ind,] <- matrix(z$grad, n, 2)[ind, ]*1.0001
  }
  if(any(is.na(grad))){
    warning(paste("IRhomogenQL/grad\n","par",par,"sigma",sig,
                  "\n fv",z$fval,"\n CC",CC,"ind",ind,"\n grad",grad),call.=FALSE)
  }
  attr(fval, "gradient") <- grad
  fval
}

IRmix2QL <- function(par, InvTimes, S0f, Rf, CL, sig, L){
  ##
  ## Inversion Recovery MRI 2 compartments (fluid/solid) with fixed fluid parameters
  ##
  ## S_{InvTime} = S0f * abs( par[1] * (1-2*exp(-InvTime*Rf)) +
  ##                        (1-par[1])*par[3]* (1-2*exp(-InvTime*par[2])) )
  ##
  n <- length(InvTimes)
  z <- .Fortran(C_irmix,
                as.double(par),
                as.double(InvTimes),
                as.double(S0f),
                as.double(Rf),
                as.integer(n),
                fval = double(n),
                grad = double(3*n))[c("fval", "grad")]
  sfval <- z$fval/sig
  fval <- CL * hg1f1(-.5, L, -sfval*sfval/2)
  CC <- CL * hg1f1(.5, L+1, -sfval*sfval/2) * sfval/2/L/sig
  grad <- matrix(CC*z$grad, n, 3)
  ind <- sfval>100
  if(any(is.na(ind))){
    warning(paste("IRmix2QL\n","par",par,"sigma",sig,
                  "\n fv",z$fval,"\n CC",CC,"ind",ind),call.=FALSE)
  }
  if(any(ind)){
    # use LS if there is no real difference, factor to keep monotonicity
    fval[ind] <- z$fval[ind]*1.0001
    grad[ind,] <- matrix(z$grad, n, 3)[ind, ]*1.0001
  }
  if(any(is.na(grad))){
    warning(paste("IRmix2QL/grad\n","par",par,"sigma",sig,
                  "\n fv",z$fval,"\n CC",CC,"ind",ind,"\n grad",grad),call.=FALSE)
  }
  attr(fval, "gradient") <- grad
  fval
}

IRmix2fvQL <- function(par, InvTimes, S0f, Rf, CL, sig, L){
  ##
  ## Inversion Recovery MRI 2 compartments (fluid/solid) with fixed fluid parameters
  ##
  ## S_{InvTime} = S0f * abs( par[1] * (1-2*exp(-InvTime*Rf)) +
  ##                        (1-par[1])*par[3]* (1-2*exp(-InvTime*par[2])) )
  ##
  n <- length(InvTimes)
  z <- .Fortran(C_irmixfv,
                as.double(par),
                as.double(InvTimes),
                as.double(S0f),
                as.double(Rf),
                as.integer(n),
                fval = double(n))[c("fval")]
  sfval <- z$fval/sig
  fval <- CL * hg1f1(-.5, L, -sfval*sfval/2)
  ind <- sfval>100
  if(any(is.na(ind))){
    warning(paste("IRmix2QL\n","par",par,"sigma",sig,
                  "\n fv",z$fval,"\n CC",fval,"ind",ind),call.=FALSE)
  }
  if(any(ind)){
    # use LS if there is no real difference, factor to keep monotonicity
    fval[ind] <- z$fval[ind]*1.0001
  }
  fval
}

IRmix5QL <- function(par, InvTimes, CL, sig, L){
  ##
  ## Inversion Recovery MRI 2 compartments (fluid/solid) with fixed fluid parameters
  ##
  ## S_{InvTime} = S0f * abs( par[1] * (1-2*exp(-InvTime*Rf)) +
  ##                        (1-par[1])*par[3]* (1-2*exp(-InvTime*par[2])) )
  ##
  n <- length(InvTimes)
  z <- .Fortran(C_irmix5,
                as.double(par),
                as.double(InvTimes),
                as.integer(n),
                fval = double(n),
                grad = double(3*n))[c("fval", "grad")]
  sfval <- z$fval/sig
  fval <- CL * hg1f1(-.5, L, -sfval*sfval/2)
  CC <- CL * hg1f1(.5, L+1, -sfval*sfval/2) * sfval/2/L/sig
  grad <- matrix(CC*z$grad, n, 3)
  ind <- sfval>100
  if(any(is.na(ind))){
    warning(paste("IRmix2QL\n","par",par,"sigma",sig,
                  "\n fv",z$fval,"\n CC",CC,"ind",ind),call.=FALSE)
  }
  if(any(ind)){
    # use LS if there is no real difference, factor to keep monotonicity
    fval[ind] <- z$fval[ind]*1.0001
    grad[ind,] <- matrix(z$grad, n, 3)[ind, ]*1.0001
  }
  if(any(is.na(grad))){
    warning(paste("IRmix2QL/grad\n","par",par,"sigma",sig,
                  "\n fv",z$fval,"\n CC",CC,"ind",ind,"\n grad",grad),call.=FALSE)
  }
  attr(fval, "gradient") <- grad
  fval
}

IRmix5fvQL <- function(par, InvTimes, CL, sig, L){
  ##
  ## Inversion Recovery MRI 2 compartments (fluid/solid) with fixed fluid parameters
  ##
  ## S_{InvTime} = S0f * abs( par[1] * (1-2*exp(-InvTime*Rf)) +
  ##                        (1-par[1])*par[3]* (1-2*exp(-InvTime*par[2])) )
  ##
  n <- length(InvTimes)
  z <- .Fortran(C_irmix5fv,
                as.double(par),
                as.double(InvTimes),
                as.integer(n),
                fval = double(n))[c("fval")]
  sfval <- z$fval/sig
  fval <- CL * hg1f1(-.5, L, -sfval*sfval/2)
  ind <- sfval>100
  if(any(is.na(ind))){
    warning(paste("IRmix2QL\n","par",par,"sigma",sig,
                  "\n fv",z$fval,"\n CC",fval,"ind",ind),call.=FALSE)
  }
  if(any(ind)){
    # use LS if there is no real difference, factor to keep monotonicity
    fval[ind] <- z$fval[ind]*1.0001
  }
  fval
}

LSIRmix2QL <- function(par, Y, InvTimes, S0f, Rf, CL, sig, L){
  fval <- IRmix2fvQL(par, InvTimes, S0f, Rf, CL, sig, L)
  kval <- (Y - fval) 
  value <- sum(kval^2)
  value
}

LSIRmix2QLgrad <- function(par, Y, InvTimes, S0f, Rf, CL, sig, L){
  fval <- IRmix2QL(par, InvTimes, S0f, Rf, CL, sig, L)
  grad <- attr(fval,"gradient")
  kval <- (Y - fval) 
  value <- sum(kval^2)
  gradient <- -2 * kval %*% grad
  gradient
}

LSIRmix5QL <- function(par, Y, InvTimes, CL, sig, L){
  fval <- IRmix5fvQL(par, InvTimes, CL, sig, L)
  kval <- (Y - fval) 
  value <- sum(kval^2)
  value
}

LSIRmix5QLgrad <- function(par, Y, InvTimes, CL, sig, L){
  fval <- IRmix2QL(par, InvTimes, CL, sig, L)
  grad <- attr(fval,"gradient")
  kval <- (Y - fval) 
  value <- sum(kval^2)
  gradient <- -2 * kval %*% grad
  gradient
}

IRmix2fixQL <- function(par, InvTimes, S0f, S0s, Rf, Rs, CL, sig, L){
  ##
  ## Inversion Recovery MRI 2 compartments (fluid/solid) with fixed fluid and solid parameters
  ##              mixture parameter only
  ##
  ## S_{InvTime} = S0f * abs( par[1] * (1-2*exp(-InvTime*Rf)) +
  ##                        (1-par[1])*par[3]* (1-2*exp(-InvTime*par[2])) )
  ##
  n <- length(InvTimes)
  z <- .Fortran(C_irmix0,
                as.double(par),
                as.double(InvTimes),
                as.double(Rs),
                as.double(S0s),
                as.double(S0f),
                as.double(Rf),
                as.integer(n),
                fval = double(n),
                grad = double(n))[c("fval", "grad")]
  sfval <- z$fval/sig
  fval <- CL * hg1f1(-.5, L, -sfval*sfval/2)
  CC <- CL * hg1f1(.5, L+1, -sfval*sfval/2) * sfval/2/L/sig
  grad <- matrix(CC*z$grad, n, 1)
  ind <- sfval>100
  if(any(is.na(ind))){
    warning(paste("IRmix2fixQL\n","par",par,"sigma",sig,
                  "\n fv",z$fval,"\n CC",CC,"ind",ind),call.=FALSE)
  }
  if(any(ind)){
    # use LS if there is no real difference, factor to keep monotonicity
    fval[ind] <- z$fval[ind]*1.0001
    grad[ind,] <- matrix(z$grad, n, 1)[ind, ]*1.0001
  }
  if(any(is.na(grad))){
    warning(paste("IRmix2fixQL/grad\n","par",par,"sigma",sig,
                  "\n fv",z$fval,"\n CC",CC,"ind",ind,"\n grad",grad),call.=FALSE)
  }
  attr(fval, "gradient") <- grad
  fval
}

File: qMRI/src/IRmodels.f
      subroutine IRfluid(th, invtime, n, fval, grad)
C
C  function values and gradients (2 parameters, S_f, R_f)
C
C  fval = th(1) * abs( 1 - 2*exp(-invtime*th(2))
C
      implicit logical (a-z)
      integer n
      double precision th(2),invtime(n),fval(n),grad(n,2)
      integer i
      double precision z1,z2,th1,th2
      th1=th(1)
      th2=th(2)
      DO i=1,n
         z1=exp(-invtime(i)*th2)
         z2=1.d0-2.d0*z1
         fval(i)=th1*abs(z2)
         grad(i,1)=abs(z2)
         grad(i,2)=2.d0*z1*invtime(i)*dsign(1.d0,z2)
      END DO
      RETURN
      END

      subroutine IRmix(th, invtime, s0, r1, n, fval, grad)
C
C  function values and gradients (3 parameters, f, R_s, S_s)
C
C  fval = abs(s0*th(1)*(1-2*exp(-invtime*t1))+(1-th(1))*th(3)*(1-2*exp(-invtime*th2)))
C
      implicit logical (a-z)
      integer n
      double precision th(3),invtime(n),s0,r1,fval(n),grad(n,3)
      integer i
      double precision z1,z2,z3,th1,th2,th3,th13,fv,vz
      th1=th(1)
      th2=th(2)
      th3=th(3)
      th13=th3*(1.d0-th1)
      DO i=1,n
         z1=s0*(1.d0-2.d0*exp(-invtime(i)*r1))
         z2=exp(-invtime(i)*th2)
         z3=1.d0-2.d0*z2
         fv=th1*z1+th13*z3
         vz=dsign(1.d0,fv)
         fval(i)=abs(fv)
         grad(i,1)=(z1-th3*z3)*vz
         grad(i,2)=(th13*2.d0*z2*invtime(i))*vz
         grad(i,3)=(1.d0-th1)*z3*vz
      END DO
      RETURN
      END

      subroutine IRmix5(th, invtime, n, fval, grad)
C
C  function values and gradients (5 parameters, f, R_s, S_s, R_f, S_f)
C
C  fval = abs(th(5)*th(1)*(1-2*exp(-invtime*th(4))+(1-th(1))*th(3)*(1-2*exp(-invtime*th(2))))
C
      implicit logical (a-z)
      integer n
      double precision th(5),invtime(n),fval(n),grad(n,5)
      integer i
      double precision z1,z2,z3,z4,z5,th1,th2,th3,th4,th5,
     1                 th13,th15,fv,vz
      th1=th(1)
      th2=th(2)
      th3=th(3)
      th4=th(4)
      th5=th(5)
      th13=th3*(1.d0-th1)
      th15=th5*th1
      DO i=1,n
         z4=exp(-invtime(i)*th4)
         z1=th(5)*(1.d0-2.d0*z4)
         z2=exp(-invtime(i)*th2)
         z3=1.d0-2.d0*z2
         z5=1.d0-2.d0*z4
         fv=th1*z1+th13*z3
         vz=dsign(1.d0,fv)
         fval(i)=abs(fv)
         grad(i,1)=(z1-th3*z3)*vz
         grad(i,2)=(th13*2.d0*z2*invtime(i))*vz
         grad(i,3)=(1.d0-th1)*z3*vz
         grad(i,4)=(th15*2.d0*z4*invtime(i))*vz
         grad(i,5)=th1*z5*vz
      END DO
      RETURN
      END

      subroutine IRmixfv(th, invtime, s0, r1, n, fval)
C
C  function values  (3 parameters, f, R_s, S_s)
C
C  fval = abs(s0*th(1)*(1-2*exp(-invtime*t1))+(1-th(1))*th(3)*(1-2*exp(-invtime*th2)))
C
      implicit logical (a-z)
      integer n
      double precision th(3),invtime(n),s0,r1,fval(n)
      integer i
      double precision z1,z2,z3,th1,th2,th3,th13,fv
      th1=th(1)
      th2=th(2)
      th3=th(3)
      th13=th3*(1.d0-th1)
      DO i=1,n
         z1=s0*(1.d0-2.d0*exp(-invtime(i)*r1))
         z2=exp(-invtime(i)*th2)
         z3=1.d0-2.d0*z2
         fv=th1*z1+th13*z3
         fval(i)=abs(fv)
      END DO
      RETURN
      END

      subroutine IRmix0(th1, invtime, th2, th3, s0, t1, n, fval, grad)
C
C  function values and gradients (1 parameters, f)
C
C  fval = abs(s0*th(1)*(1-2*exp(-invtime*t1))+(1-th(1))*th(3)*(1-2*exp(-invtime*th(2))))
C
      implicit logical (a-z)
      integer n
      double precision th1,invtime(n),s0,t1,fval(n),grad(n),th2,th3
      integer i
      double precision z1,z3,th13,fv
      th13=th3*(1.d0-th1)
      DO i=1,n
         z1=s0*(1.d0-2.d0*exp(-invtime(i)*t1))
         z3=(1.d0-2.d0*exp(-invtime(i)*th2))
         fv=th1*z1+th13*z3
         fval(i)=abs(fv)
         grad(i)=(z1-th3*z3)*dsign(1.d0,fv)
      END DO
      RETURN
      END

      subroutine IRmix5fv(th, invtime, n, fval)
C
C  function values and gradients (5 parameters, f, R_s, S_s, R_f, S_f)
C
C  fval = abs(th(5)*th(1)*(1-2*exp(-invtime*th(4))+(1-th(1))*th(3)*(1-2*exp(-invtime*th(2))))
C
      implicit logical (a-z)
      integer n
      double precision th(5),invtime(n),fval(n)
      integer i
      double precision z1,z2,z3,z4,z5,th1,th2,th3,th4,th5,
     1                 th13,th15,fv
      th1=th(1)
      th2=th(2)
      th3=th(3)
      th4=th(4)
      th5=th(5)
      th13=th3*(1.d0-th1)
      th15=th5*th1
      DO i=1,n
         z4=exp(-invtime(i)*th4)
         z1=th(5)*(1.d0-2.d0*z4)
         z2=exp(-invtime(i)*th2)
         z3=1.d0-2.d0*z2
         z5=1.d0-2.d0*z4
         fv=th1*z1+th13*z3
         fval(i)=abs(fv)
      END DO
      RETURN
      END

